<h2> Jupyter Homework Week 4 --- Math 032, Summer 2024 </h2>

<img src="images/empirical-rule.png" style="float: right; width: 30%">



<h2> Simulating means and variances </h2>

<h4> Part 1: Introduction </h4>

In this notebook, we'll study the normal distribution and learn how to sample from it. We will empirically verify the [68-95-99.7 rule](https://en.wikipedia.org/wiki/68–95–99.7_rule), which states that, if we sample a large number of times from a [normal distribution](https://en.wikipedia.org/wiki/Normal_distribution), we should expect about $68$% of samples to be within *one* standard deviation of the mean, $95$% to be within *two* standard deviations, and $99.7$% to be within *three* standard deviations. Then, we will sample from a standard normal distribution (mean $0$ and variance $1$) to estimate the means and variances of two other distributions that are derived from the normal distribution.

<h4> Part 2: Generating data and Calculating Quantiles </h4>

To begin, let's use scipy's stats package to find the percentiles (or quantiles) of the standard normal distribution and compare it to the table in our textbook (Table B.1 on page 432).

In [1]:
# import the requisite package
import numpy as np
from scipy.stats import norm

Here is an example of how to use `norm` to compute quantiles:

In [2]:
# Define the mean and standard deviation
mean = 0
std_dev = 1

# Create a normal distribution object
normal_dist = norm(loc=mean, scale=std_dev)

# Calculate quantiles
quantile_95 = normal_dist.ppf(0.95)
quantile_80 = normal_dist.ppf(0.80)

print("95th percentile (quantile):", quantile_95)
print("80th percentile (quantile):", quantile_80)

95th percentile (quantile): 1.6448536269514722
80th percentile (quantile): 0.8416212335729143


This means that, on the graph above, $95$% of the area is to the left of $x=1.64485$, and $80$% is to the left of $0.8416$. We can also generate samples using the following syntax:

In [3]:
sample_size = 10000
samples = normal_dist.rvs(size=sample_size)
print("The first five samples are ",samples[:5])

The first five samples are  [ 0.18313712 -0.75709303 -0.79445668  1.14406146  1.27052271]


Let's continue: we can compute means and variances by directly computing averages.

In [4]:
print("The average of our samples is",samples.mean())
print("The variance of our samples is",samples.var())

The average of our samples is -0.015926105745871194
The variance of our samples is 0.9954333496683612


Notice how close these are to the true values $\mu = 0$ and $\sigma^2 = 1$!

Let's suppose $Z$ is a *standard normal* random variable, so that $Z$ has pdf

$$f(x) = \frac{1}{\sqrt{2 \pi}} e^{-x^2/2}, \ \ \ \ -\infty <x < \infty .$$
If we want to calculate exact values of the cumulative distribution function (cdf), we can use `norm.cdf`.  Here are two ways to do this. Note that we have *already defined* `normal_dist` above.

In [5]:
# Calculate the probability P(Z <= 1).
print(normal_dist.cdf(1))
# Calculate the probability P(-4 <= Z <= 2).
prob = normal_dist.cdf(2) - normal_dist.cdf(-4)
print(prob)

0.8413447460685429
0.9772181968099877


### Part 3: Questions

#### Question 1:  

Estimate a threshold $\alpha$ for which $P(Z \le \alpha) = 0.75$ using `normal_dist.ppf`.

In [6]:
# Put your answer to question 1 here

alpha = normal_dist.ppf(0.75) # SOLUTION
print(alpha)

0.6744897501960817


In [14]:
round(alpha, 4) == 0.6745

True

#### Question 2:

A standard rule of thumb described above is the [68-95-99.7](https://en.wikipedia.org/wiki/68–95–99.7_rule) rule, which refers to the probabilities of $P(-1 \le Z \le 1)$, $P(-2 \le Z \le 2)$, and $P(-3 \le Z \le 3)$ for a random variable $Z \sim N(0, 1)$. 

Calculate these probabilities using the function `norm.cdf`. You can see an example above. Your answer should be *very close* to the true values, because you are not doing any sampling.

In [15]:
# Put your answer to question 2 here

prob_within_one = norm.cdf(1)-norm.cdf(-1) # SOLUTION
prob_within_two = norm.cdf(2)-norm.cdf(-2) # SOLUTION
prob_within_three = norm.cdf(3)-norm.cdf(-3) # SOLUTION

print(prob_within_one,prob_within_two,prob_within_three)

0.6826894921370859 0.9544997361036416 0.9973002039367398


In [9]:
assert round(prob_within_one,4) == round(norm.cdf(1)-norm.cdf(-1),4)
assert round(prob_within_two,4) == round(norm.cdf(2)-norm.cdf(-2),4)
assert round(prob_within_three,4) == round(norm.cdf(3)-norm.cdf(-3),4)

#### Question 3:

Many important distributions arise by *applying a function* to normally distributed data. Two of these are the  [chi-squared distribution](https://en.wikipedia.org/wiki/Chi-squared_distribution), which is of the form $X_1^2+X_2^2+...+X_n^2$, where the $X_i$ are normally distributed, and the [folded normal distribution](https://en.wikipedia.org/wiki/Folded_normal_distribution), which is the absolute value $|Z|$ of a standard normal random variable.

Suppose $Z$ is standard normal and let $X = Z^2$, $Y=|Z|$. **Estimate the mean and variance** of $X$ and $Y$ *by sampling* from a standard normal distribution. You can use `np.random.normal()` to create samples of $Z$.

In [10]:
# Put your answer to question 3 here

def estimate_mean_and_variance_X():
    # sample chi_squared and return the sample mean and variance
    # Take a large number of samples (e.g. 100000)
    num_samples = 100000 # SOLUTION
    # Sample From a Standard Normal Distribution
    Z = np.random.normal(size = num_samples) # SOLUTION
    # Create samples of chi-square from normal_samples
    X = Z**2 # SOLUTION
    return X.mean(),X.var()

def estimate_mean_and_variance_Y():
    # sample from folded normal and return the sample mean and variance
    # Take a large number of samples (e.g. 100000)
    num_samples = 100000 # SOLUTION
    # Sample From a Standard Normal Distribution
    Z = np.random.normal(size = num_samples) # SOLUTION
    # Create samples of folded normal
    Y = abs(Z) # SOLUTION
    return Y.mean(),Y.var()


print("The chi squared distribution has mean and variance: ",estimate_mean_and_variance_X())
print("The folded normal distribution has mean and variance: ",estimate_mean_and_variance_Y())

The chi squared distribution has mean and variance:  (0.9947804212159971, 1.9933211377595865)
The folded normal distribution has mean and variance:  (0.7976654912848884, 0.36214452736924946)


In [11]:
samples1 = [estimate_mean_and_variance_X() for _ in range(100)]
samples2 = [estimate_mean_and_variance_Y() for _ in range(100)]
means1 = sum([m[0] for m in samples1])/100
means2 = sum([m[0] for m in samples2])/100
var1 = sum([m[1] for m in samples1])/100
var2 = sum([m[1] for m in samples2])/100
assert abs(1 - round(means1,3)) <= 0.05
assert abs(2 - round(var1,3)) <= 0.05
assert abs(np.sqrt(2/np.pi) - round(means2,3)) <= 0.05
assert abs(1-2/np.pi - round(var2,3)) <= 0.05